consensus/src/chained_bft/block_storage/block_store.rs

```diff
@@ -209,21 +218,30 @@ impl<T: Payload> BlockStore<T> {
                 return Err(e);
             }
         };
-        let compute_res = self
-            .state_computer
-            .compute(
-                parent_id,
-                block.id(),
-                block.payload().unwrap_or(&T::default()),
-            )
-            .await
-            .with_context(|e| format!("Execution failure for block {}: {:?}", block, e))?;
-
-        self.storage
-            .save_tree(vec![block.clone()], vec![])
-            .with_context(|e| format!("Insert block failed with {:?} when saving block", e))?;
-        let new_block = ExecutedBlock::new(block, compute_res);
-        self.inner.write().unwrap().insert_block(new_block)
+        // Reconfiguration rule - if a block is a child of reconfiguration, it needs to be empty
+        // So we roll over the executed state until it's committed and we start new epoch.
+        let parent_state = parent_block.compute_result();
+        let compute_res = if parent_state.executed_state.validators.is_some() {
+            ensure!(
+                block.payload().filter(|p| **p != T::default()).is_none(),
+                "Reconfiguration suffix should not carry payload"
+            );
+            StateComputeResult {
+                executed_state: parent_state.executed_state.clone(),
+                compute_status: vec![],
+            }
+        } else {
+            // We still execute NIL block for the block prologue transaction.
```

Comment:
Hmm I don't quite understand this comment: why do we talk about NIL blocks here?

Reply:
let me rephrase it, NIL blocks don't have payload at all, but we still pass a T::default() to executor because we'll inject a special transaction.